File: shared/src/types/incident.ts
export enum IncidentStatus {
  OPEN = 'OPEN',
  IN_PROGRESS = 'IN_PROGRESS',
  RESOLVED = 'RESOLVED',
  CLOSED = 'CLOSED',
  CANCELLED = 'CANCELLED'
}

export enum IncidentPriority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL'
}

export enum IncidentSeverity {
  MINOR = 'MINOR',
  MODERATE = 'MODERATE',
  MAJOR = 'MAJOR',
  CRITICAL = 'CRITICAL'
}

export enum UserRole {
  ADMIN = 'ADMIN',
  MANAGER = 'MANAGER',
  RESPONDER = 'RESPONDER',
  VIEWER = 'VIEWER'
}

export interface User {
  id: string;
  email: string;
  firstName: string;
  lastName: string;
  role: UserRole;
  avatar?: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface Incident {
  id: string;
  title: string;
  description: string;
  status: IncidentStatus;
  priority: IncidentPriority;
  severity: IncidentSeverity;
  assigneeId?: string;
  assignee?: User;
  reporterId: string;
  reporter: User;
  tags: string[];
  createdAt: Date;
  updatedAt: Date;
  resolvedAt?: Date;
  closedAt?: Date;
}

export interface IncidentUpdate {
  id: string;
  incidentId: string;
  userId: string;
  user: User;
  content: string;
  type: 'COMMENT' | 'STATUS_CHANGE' | 'ASSIGNMENT' | 'PRIORITY_CHANGE';
  metadata?: Record<string, any>;
  createdAt: Date;
}

export interface IncidentAttachment {
  id: string;
  incidentId: string;
  fileName: string;
  fileSize: number;
  mimeType: string;
  uploadedBy: string;
  uploadedAt: Date;
  url: string;
}

export interface CreateIncidentRequest {
  title: string;
  description: string;
  priority: IncidentPriority;
  severity: IncidentSeverity;
  assigneeId?: string;
  tags?: string[];
}

export interface UpdateIncidentRequest {
  title?: string;
  description?: string;
  status?: IncidentStatus;
  priority?: IncidentPriority;
  severity?: IncidentSeverity;
  assigneeId?: string;
  tags?: string[];
}

export interface IncidentFilters {
  status?: IncidentStatus[];
  priority?: IncidentPriority[];
  severity?: IncidentSeverity[];
  assigneeId?: string;
  reporterId?: string;
  tags?: string[];
  dateFrom?: Date;
  dateTo?: Date;
  search?: string;
} 

File: backend/internal/models/incident.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type IncidentStatus string
type IncidentPriority string
type IncidentSeverity string

const (
	StatusOpen       IncidentStatus = "OPEN"
	StatusInProgress IncidentStatus = "IN_PROGRESS"
	StatusResolved   IncidentStatus = "RESOLVED"
	StatusClosed     IncidentStatus = "CLOSED"
	StatusCancelled  IncidentStatus = "CANCELLED"
)

const (
	PriorityLow      IncidentPriority = "LOW"
	PriorityMedium   IncidentPriority = "MEDIUM"
	PriorityHigh     IncidentPriority = "HIGH"
	PriorityCritical IncidentPriority = "CRITICAL"
)

const (
	SeverityMinor    IncidentSeverity = "MINOR"
	SeverityModerate IncidentSeverity = "MODERATE"
	SeverityMajor    IncidentSeverity = "MAJOR"
	SeverityCritical IncidentSeverity = "CRITICAL"
)

type Incident struct {
	ID          uint             `json:"id" gorm:"primaryKey"`
	Title       string           `json:"title" gorm:"not null"`
	Description string           `json:"description" gorm:"type:text;not null"`
	Status      IncidentStatus   `json:"status" gorm:"not null;default:'OPEN'"`
	Priority    IncidentPriority `json:"priority" gorm:"not null"`
	Severity    IncidentSeverity `json:"severity" gorm:"not null"`
	AssigneeID  *uint            `json:"assigneeId"`
	Assignee    *User            `json:"assignee" gorm:"foreignKey:AssigneeID"`
	ReporterID  uint             `json:"reporterId" gorm:"not null"`
	Reporter    User             `json:"reporter" gorm:"foreignKey:ReporterID"`
	Tags        []string         `json:"tags" gorm:"type:text[]"`
	ResolvedAt  *time.Time       `json:"resolvedAt"`
	ClosedAt    *time.Time       `json:"closedAt"`
	CreatedAt   time.Time        `json:"createdAt"`
	UpdatedAt   time.Time        `json:"updatedAt"`
	DeletedAt   gorm.DeletedAt   `json:"-" gorm:"index"`
}

func (Incident) TableName() string {
	return "incidents"
}
